### ISSUE TITLE
Course: Gen AI Foundational Models for NLP & Language Understanding - Module 1

### ISSUE THREAD

**Issue by niyontezejado** [niyontezejado/GenAI#3]
#  Module 1: Fundamentals of Language Understanding

## A. Language Understanding with Neural Networks

**1. What to learn:** describe fundamentals of language understanding, explain the NLP models and techniques.

**2. Converting Words to Features**: 

- One-hot-encoding: the way to use in converting categorical features into feature vectors.

- The bag of words representation portrays a document as the aggregate or average of one-hot encoded vectors.

- When you feed a bag-of-words vector to a neural network's hidden layer, the output is the sum of the embeddings.

- The embedding and embeddingBag classes implement embedding and embedding bags in PyTorch.

**3. Document Categorization Prediction with Torchtext**:

- Document classifier seamlessly categorizes articles by analyzing the text content.

- Neural network is a mathematical function consisting of a sequence of matrix multiplications with a variety of other functions.

- Argmax function identifies the index of the highest logit value, corresponding to the most likely class.

- Hyperparameters are externally set configurations of a neural network.

- Prediction function:
        - Works on real text that starts by taking in tokenized text.
        - Processes the text through the pipeline, and the model predicts the category.

**4. Document Categorization Training with Torchtext**:

- Neural Network functions via matrix and vector operations called learnable parameters.
- In neural network training:
    - Learnable parameters are fine-tuned to enhance model performance.
    -  Process is steered by the loss function, which serves as a measure of accuracy.
- Prediction function:
    - Works on real text that starts by taking in tokenized text. 
    - Processes the text through the pipeline; the model predicts the category.
- Cross-entropy is used to find the best parameters.
- For unknown distribution, you can estimate it by averaging the function applied to a set of samples. This is known as Monte Carlo sampling.
- Optimization is used to minimize the loss.
- Three subsets of partitioned data set are: training data, validation data, and test data.

**5. Training the model in PyTorch**:

- Training data is split into training and validation, and then data loaders are set up for training, validation and testing.
- Batch size specifies the sample count for gradient approximation
- Data shuffling promotes better optimization
- When difining model, init_weights helps with optimization
- To train your loop:
    - Iterate over each epoch.
    - Set model to training mode and calculate the total loss.
    -  Devide the data set into batches.
    - Perform gradient descent.
    - Update loss after each batch is processed.

## B. N-Gram Model

<img width="1757" height="725" alt="Image" src="https://github.com/user-attachments/assets/28ce5ccf-2dc6-4ca6-bcc9-c38393db493e" />

-----

<img width="1130" height="476" alt="Image" src="https://github.com/user-attachments/assets/49fe07c6-bd23-411e-b200-7792709861a1" />
